From a80ad59f02e381d3c06a5e672189421fc35fc017 Mon Sep 17 00:00:00 2001
From: Jianhui Zhao <zhaojh329@gmail.com>
Date: Mon, 11 Aug 2025 18:23:05 +0800
Subject: [PATCH] fix(ssl): Set the server_name before the handshake

Signed-off-by: Jianhui Zhao <zhaojh329@gmail.com>
---
 http/client.lua | 1 +
 ssl.lua         | 8 ++++----
 2 files changed, 5 insertions(+), 4 deletions(-)

diff --git a/http/client.lua b/http/client.lua
index fd895ab..9d7a2a0 100644
--- a/http/client.lua
+++ b/http/client.lua
@@ -407,6 +407,7 @@ function methods:request(method, url, body, opts)
     for _, address in ipairs(addresses) do
         if scheme_info.use_ssl then
             local ssl = require 'eco.ssl'
+            opts.server_name = host
             sock, err = ssl.connect(address, port, opts)
         else
             sock, err = socket.connect_tcp(address, port, opts)
diff --git a/ssl.lua b/ssl.lua
index a4310d4..6bd6276 100644
--- a/ssl.lua
+++ b/ssl.lua
@@ -31,10 +31,6 @@ end
 
 local cli_methods = {}
 
-function cli_methods:set_server_name(name)
-    return self.ssock:set_server_name(name)
-end
-
 function cli_methods:send(data)
     local mutex = self.mutex
 
@@ -251,6 +247,10 @@ function M.connect(ipaddr, port, options)
 
     local ssock = ctx:new(sock:getfd(), options.insecure)
 
+    if options.server_name then
+        ssock:set_server_name(options.server_name)
+    end
+
     local ok, err = ssock:handshake()
     if not ok then
         ssock:free()
-- 
2.34.1

