From b3a3844efc6b753c093aba40d9fb57d04053d872 Mon Sep 17 00:00:00 2001
From: Jianhui Zhao <zhaojh329@gmail.com>
Date: Sat, 10 May 2025 10:05:27 +0800
Subject: [PATCH] fix: Fix set hook

Signed-off-by: Jianhui Zhao <zhaojh329@gmail.com>
---
 eco.c | 24 +++++++++++++++---------
 1 file changed, 15 insertions(+), 9 deletions(-)

diff --git a/eco.c b/eco.c
index 14dfe63..2fb31d8 100644
--- a/eco.c
+++ b/eco.c
@@ -82,8 +82,6 @@ static int eco_all(lua_State *L)
 
 static int eco_run(lua_State *L)
 {
-    const char *prehook = "return function(t,f,...) \
-        local hook, mask = debug.gethook(t) if hook then debug.sethook(hook, mask) end f(...) end";
     int narg = lua_gettop(L);
     lua_State *co;
 
@@ -110,16 +108,24 @@ static int eco_run(lua_State *L)
     lua_rawset(L, -3);  /* ctx_env[co] = tmr_ptr(NULL) */
     lua_pop(L, 1);
 
-    if (luaL_dostring(co, prehook))
-        return lua_error(L);
+    lua_getglobal(L, "debug");
+    lua_getfield(L, -1, "gethook");
+    lua_remove(L, -2);
+
+    lua_call(L, 0, 3);
+
+    lua_getglobal(L, "debug");
+    lua_getfield(L, -1, "sethook");
+    lua_remove(L, -2);
+
+    lua_pushthread(co);
+    lua_xmove(co, L, 1);
 
-    /* push parent thread to top of new thread */
-    lua_pushthread(L);
-    lua_xmove(L, co, 1);
+    lua_rotate(L, -5, 2);
 
-    lua_rotate(co, 1, 2);
+    lua_call(L, 4, 0);
 
-    eco_resume(L, co, narg + 1);
+    eco_resume(L, co, narg - 1);
 
     return 0;
 }
-- 
2.34.1

