Index: b/socket.lua
===================================================================
--- a/socket.lua	2024-12-26 11:44:23.000000000 +0800
+++ b/socket.lua	2025-04-16 17:56:50.027366825 +0800
@@ -182,6 +182,7 @@ function M.socket(family, domain, protoc
         sock:setoption('bindtodevice', options.device)
     end
 
+    sock:setoption('sndbuf', 1400)    
     return setmetatable(o, metatable)
 end
 
