From 3b784010f169b13984bfaf8dd84fef13cabbd8b6 Mon Sep 17 00:00:00 2001
From: Jianhui Zhao <zhaojh329@gmail.com>
Date: Thu, 23 Oct 2025 17:34:46 +0800
Subject: [PATCH 3/7] fix(socket): return closed status instead of discarding
 it

Signed-off-by: Jianhui Zhao <zhaojh329@gmail.com>
---
 socket.lua | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/socket.lua b/socket.lua
index c9c109a..0765fe2 100644
--- a/socket.lua
+++ b/socket.lua
@@ -22,7 +22,7 @@ function methods:close()
 end
 
 function methods:closed()
-    self.sock:closed()
+    return self.sock:closed()
 end
 
 function methods:setoption(name, value)
-- 
2.34.1

