Index: lua-eco-d546a9b7/eco.c
===================================================================
--- lua-eco-d546a9b7.orig/eco.c
+++ lua-eco-d546a9b7/eco.c
@@ -93,10 +93,21 @@ static int eco_count(lua_State *L)
     return 1;
 }
 
+static void sigint_handler(struct ev_loop *loop, struct ev_signal *w, int e)
+{
+	printf("Got SIGINT. Now quit\n");
+	ev_break(loop, EVBREAK_ALL);
+}
+
 int luaopen_eco(lua_State *L)
 {
     struct ev_loop *loop = EV_DEFAULT;
     struct eco_context *ctx;
+	static struct ev_signal sig;
+
+
+	ev_signal_init(&sig, sigint_handler, SIGINT);
+	ev_signal_start(loop, &sig);
 
     lua_pushstring(L, ECO_CONTEXT_KEY);
     ctx = lua_newuserdata(L, sizeof(struct eco_context));
