--- a/dns.lua
+++ b/dns.lua
@@ -4,6 +4,7 @@
 -- Referenced from https://github.com/openresty/lua-resty-dns/blob/master/lib/resty/dns/resolver.lua
 
 local socket = require 'eco.socket'
+local file = require 'eco.file'
 local bit = require 'eco.bit'
 
 local str_char = string.char
@@ -60,6 +61,10 @@ local function parse_resolvconf()
     local nameservers = {}
     local conf = {}
 
+    if not file.access('/etc/resolv.conf') then
+        return nil, 'no /etc/resolv.conf'
+    end
+
     for line in io.lines('/etc/resolv.conf') do
         if line:match('search') then
             local search = line:match('search%s+(.+)')
@@ -562,14 +567,17 @@ function M.query(qname, opts)
         nameservers[#nameservers + 1] = { host, port, socket.is_ipv6_address(host) }
     end
 
-    local resolvconf = parse_resolvconf()
+    local resolvconf, err = parse_resolvconf()
+    if not resolvconf then
+        return nil, err
+    end
 
     for _, nameserver in ipairs(resolvconf.nameservers) do
         nameservers[#nameservers + 1] = nameserver
     end
 
     if #nameservers < 1 then
-        error('not found valid nameservers')
+        return nil, 'not found valid nameservers'
     end
 
     if not qname:match('%.') and resolvconf.search then
