From 360272af92b21a38ef2f41934b03982dcc6bc19b Mon Sep 17 00:00:00 2001
From: Jianhui Zhao <zhaojh329@gmail.com>
Date: Sat, 27 Dec 2025 17:34:35 +0800
Subject: [PATCH] feat(nl80211): add Lua binding for 320MHz channel width
 constant

Signed-off-by: Jianhui Zhao <zhaojh329@gmail.com>
---
 nl80211.c   | 1 +
 nl80211.lua | 3 ++-
 2 files changed, 3 insertions(+), 1 deletion(-)

diff --git a/nl80211.c b/nl80211.c
index 9e4f4cc..8600d0e 100644
--- a/nl80211.c
+++ b/nl80211.c
@@ -318,6 +318,7 @@ int luaopen_eco_core_nl80211(lua_State *L)
     lua_add_constant(L, "CHAN_WIDTH_160", NL80211_CHAN_WIDTH_160);
     lua_add_constant(L, "CHAN_WIDTH_5", NL80211_CHAN_WIDTH_5);
     lua_add_constant(L, "CHAN_WIDTH_10", NL80211_CHAN_WIDTH_10);
+    lua_add_constant(L, "CHAN_WIDTH_320", NL80211_CHAN_WIDTH_320);
 
     lua_add_constant(L, "CHAN_NO_HT", NL80211_CHAN_NO_HT);
     lua_add_constant(L, "CHAN_HT20", NL80211_CHAN_HT20);
diff --git a/nl80211.lua b/nl80211.lua
index 6ca485d..a737509 100644
--- a/nl80211.lua
+++ b/nl80211.lua
@@ -136,7 +136,8 @@ local width_name = {
     [nl80211.CHAN_WIDTH_40] = '40 MHz',
     [nl80211.CHAN_WIDTH_80] = '80 MHz',
     [nl80211.CHAN_WIDTH_80P80] = '80+80 MHz',
-    [nl80211.CHAN_WIDTH_160] = '160 MHz'
+    [nl80211.CHAN_WIDTH_160] = '160 MHz',
+    [nl80211.CHAN_WIDTH_320] = '320 MHz'
 }
 
 local channel_type_name = {
-- 
2.34.1

