Index: ksmbd-3.3.7/transport_rdma.c
===================================================================
--- ksmbd-3.3.7.orig/transport_rdma.c
+++ ksmbd-3.3.7/transport_rdma.c
@@ -997,8 +997,8 @@ static int smb_direct_create_header(stru
 	int ret;
 
 	sendmsg = smb_direct_alloc_sendmsg(t);
-	if (!sendmsg)
-		return -ENOMEM;
+	if (IS_ERR(sendmsg))
+		return PTR_ERR(sendmsg);
 
 	/* Fill in the packet header */
 	packet = (struct smb_direct_data_transfer *)sendmsg->packet;
