Index: ksmbd-3.3.7/mgmt/user_session.c
===================================================================
--- ksmbd-3.3.7.orig/mgmt/user_session.c
+++ ksmbd-3.3.7/mgmt/user_session.c
@@ -316,9 +316,9 @@ static struct ksmbd_session *__session_c
 		goto error;
 
 	if (protocol == CIFDS_SESSION_FLAG_SMB2) {
-		down_read(&sessions_table_lock);
+		down_write(&sessions_table_lock);
 		hash_add(sessions_table, &sess->hlist, sess->id);
-		up_read(&sessions_table_lock);
+		up_write(&sessions_table_lock);
 	}
 	return sess;
 
