Index: ot-br-posix-2022-11-08/src/agent/application.cpp
===================================================================
--- ot-br-posix-2022-11-08.orig/src/agent/application.cpp
+++ ot-br-posix-2022-11-08/src/agent/application.cpp
@@ -41,6 +41,7 @@
 #include "common/code_utils.hpp"
 #include "common/mainloop_manager.hpp"
 #include "utils/infra_link_selector.hpp"
+#include "utils/uci_config.hpp"
 
 namespace otbr {
 
@@ -102,6 +103,8 @@ void Application::Init(void)
 #if OTBR_ENABLE_VENDOR_SERVER
     mVendorServer.Init();
 #endif
+
+    otbr::Utils::otConfigInitFromUCI(mNcp.GetInstance());
 }
 
 void Application::Deinit(void)
