Index: ot-br-posix-2022-11-08-r4.2.3/third_party/openthread/repo/src/posix/platform/udp.cpp
===================================================================
--- ot-br-posix-2022-11-08-r4.2.3.orig/third_party/openthread/repo/src/posix/platform/udp.cpp
+++ ot-br-posix-2022-11-08-r4.2.3/third_party/openthread/repo/src/posix/platform/udp.cpp
@@ -296,6 +296,13 @@ exit:
     if (error == OT_ERROR_FAILED)
     {
         otLogCritPlat("Failed to bind UDP socket: %s", strerror(errno));
+        if (errno == EINVAL)
+        {
+            char str[INET6_ADDRSTRLEN];
+            inet_ntop(AF_INET6, &(aUdpSocket->mSockName.mAddress), str, INET6_ADDRSTRLEN);
+            otLogCritPlat("Got Invalid argument, addr=%s, port=%d", str, htons(aUdpSocket->mSockName.mPort));
+            abort();
+        }
     }
 
     return error;
