Index: ot-br-posix-2022-07-11/third_party/openthread/repo/src/core/net/ip6_mpl.cpp
===================================================================
--- ot-br-posix-2022-07-11.orig/third_party/openthread/repo/src/core/net/ip6_mpl.cpp
+++ ot-br-posix-2022-07-11/third_party/openthread/repo/src/core/net/ip6_mpl.cpp
@@ -191,6 +191,7 @@ Error Mpl::UpdateSeedSet(uint16_t aSeedI
             if (aSequence == mSeedSet[i].mSequence)
             {
                 // already received, drop message
+                mSeedSet[i].mLifetime = kSeedEntryLifetime;
                 ExitNow(error = kErrorDrop);
             }
             else if (insert == nullptr && SerialNumber::IsLess(aSequence, mSeedSet[i].mSequence))
