diff --git a/src/dhcp.c b/src/dhcp.c
index 6775fb6..2989b9a 100644
--- a/src/dhcp.c
+++ b/src/dhcp.c
@@ -322,7 +322,7 @@ void dhcp_packet(time_t now, int pxe_fd)
       else
 	{
 	  if (iface_check(AF_INET, NULL, ifr.ifr_name, NULL))
-	    my_syslog(MS_DHCP | LOG_WARNING, _("DHCP packet received on %s which has no address"), ifr.ifr_name);
+	    my_syslog(MS_DHCP | LOG_DEBUG, _("DHCP packet received on %s which has no address"), ifr.ifr_name);
 	  return;
 	}
       
diff --git a/src/dnsmasq.h b/src/dnsmasq.h
index 1111111..e6f5287 100644
--- a/src/dnsmasq.h
+++ b/src/dnsmasq.h
@@ -1207,7 +1207,7 @@ extern struct daemon {
   int max_logs;  /* queue limit */
   int randport_limit; /* Maximum number of source ports for query. */
   int cachesize, ftabsize;
-  int port, query_port, min_port, max_port;
+  int port, query_port, min_port, max_port, log_level;
   unsigned long local_ttl, neg_ttl, max_ttl, min_cache_ttl, max_cache_ttl, auth_ttl, dhcp_ttl, use_dhcp_ttl;
   char *dns_client_id;
   u32 umbrella_org;
diff --git a/src/log.c b/src/log.c
index b46fef8..7e9ab70 100644
--- a/src/log.c
+++ b/src/log.c
@@ -317,6 +317,9 @@ void my_syslog(int priority, const char *format, ...)
   priority &= LOG_PRIMASK;
 #endif
 
+  if (daemon->log_level > 0 && priority > daemon->log_level)
+    return;
+
   if (echo_stderr) 
     {
       fprintf(stderr, "dnsmasq%s: ", func);
diff --git a/src/option.c b/src/option.c
index 3333333..aa5bb96 100644
--- a/src/option.c
+++ b/src/option.c
@@ -197,7 +197,8 @@ struct myoption {
 #define LOPT_DO_ENCODE     388
 #define LOPT_LEASEQUERY    389
 #define LOPT_SPLIT_RELAY   390
-#define LOPT_DHCP_MODE     503
+#define LOPT_DHCP_MODE     503
+#define LOPT_LOG_LEVEL     504
 
 #ifdef HAVE_GETOPT_LONG
 static const struct option opts[] =  
@@ -390,6 +391,7 @@ static const struct myoption opts[] =
     { "dhcp-ignore-clid", 0, 0,  LOPT_IGNORE_CLID },
     { "dynamic-host", 1, 0, LOPT_DYNHOST },
     { "dhcp-mode", 1, 0, LOPT_DHCP_MODE },
+    { "log-level", 1, 0, LOPT_LOG_LEVEL },
     { "log-debug", 0, 0, LOPT_LOG_DEBUG },
     { "umbrella", 2, 0, LOPT_UMBRELLA },
     { "quiet-tftp", 0, 0, LOPT_QUIET_TFTP },
@@ -600,6 +602,7 @@ static struct {
   { LOPT_DUMPMASK, ARG_ONE, "<hex>", gettext_noop("Mask which packets to dump."), NULL },
   { LOPT_SCRIPT_TIME, OPT_LEASE_RENEW, NULL, gettext_noop("Call dhcp-script when lease expiry changes."), NULL },
   { LOPT_DHCP_MODE, ARG_ONE, "<mode>", gettext_noop("DHCP mode router or passthrough (GL)."), NULL },
+  { LOPT_LOG_LEVEL, ARG_ONE, "<level>", gettext_noop("Configure maximum syslog priority for dnsmasq (GL)."), NULL },
   { LOPT_UMBRELLA, ARG_ONE, "[=<optspec>]", gettext_noop("Send Cisco Umbrella identifiers including remote IP."), NULL },
   { LOPT_QUIET_TFTP, OPT_QUIET_TFTP, NULL, gettext_noop("Do not log routine TFTP."), NULL },
   { LOPT_NORR, OPT_NORR, NULL, gettext_noop("Suppress round-robin ordering of DNS records."), NULL },
@@ -2375,6 +2378,11 @@ static int one_opt(int option, char *arg, char *errstr, char *gen_err, int comma
 
     case LOPT_DHCP_MODE:  /* --dhcp-mode */
       daemon->mode = opt_string_alloc(arg);
+      break;
+
+    case LOPT_LOG_LEVEL:  /* --log-level */
+      if (!atoi_check16(arg, &daemon->log_level))
+	ret_err(gen_err);
       break;
       
 #ifdef HAVE_DHCP      
@@ -5976,6 +5984,7 @@ void read_opts(int argc, char **argv, char *compile_opts)
   daemon->soa_retry = SOA_RETRY;
   daemon->soa_expiry = SOA_EXPIRY;
   daemon->randport_limit = 1;
+  daemon->log_level = LOG_INFO;
   daemon->host_index = SRC_AH;
   daemon->max_procs = MAX_PROCS;
 #ifdef HAVE_DUMPFILE
