Index: wireguard-linux-compat-1.0.20200611/src/compat/compat.h
===================================================================
--- wireguard-linux-compat-1.0.20200611.orig/src/compat/compat.h
+++ wireguard-linux-compat-1.0.20200611/src/compat/compat.h
@@ -668,12 +668,12 @@ struct __compat_dummy_container { char d
 #endif
 
 #if LINUX_VERSION_CODE < KERNEL_VERSION(4, 13, 0) && !defined(ISRHEL7) && !defined(ISOPENSUSE15)
-static inline void *skb_put_data(struct sk_buff *skb, const void *data, unsigned int len)
+/*static inline void *skb_put_data(struct sk_buff *skb, const void *data, unsigned int len)
 {
 	void *tmp = skb_put(skb, len);
 	memcpy(tmp, data, len);
 	return tmp;
-}
+}*/
 #endif
 
 #if LINUX_VERSION_CODE < KERNEL_VERSION(3, 19, 0) && !defined(ISRHEL7)
