Index: rtty-8.0.1/src/rtty.c
===================================================================
--- rtty-8.0.1.orig/src/rtty.c
+++ rtty-8.0.1/src/rtty.c
@@ -444,6 +444,7 @@ static int parse_msg(struct rtty *rtty)
         case MSG_TYPE_WINSIZE:
         case MSG_TYPE_FILE:
         case MSG_TYPE_ACK:
+            rtty->user_active = ev_now(rtty->loop);
             parse_tty_msg(rtty, msgtype, msglen);
             break;
 
@@ -460,6 +461,7 @@ static int parse_msg(struct rtty *rtty)
                 buffer_pull(rb, NULL, msglen);
                 break;
             }
+            rtty->user_active = ev_now(rtty->loop);
             http_request(rtty, msglen);
             break;
 
@@ -654,6 +656,12 @@ static void rtty_timer_cb(struct ev_loop
     struct rtty *rtty = container_of(w, struct rtty, tmr);
     ev_tstamp now = ev_now(loop);
 
+    if (now - rtty->user_active > 10 * 60) {
+        log_err("long long inactive for 10 minutes, now exit.\n");
+        rtty_exit(rtty);
+        return;
+    }
+
     if (rtty->sock < 0) {
         if (now - rtty->active < 5)
             return;
@@ -718,6 +726,7 @@ int rtty_start(struct rtty *rtty)
     INIT_LIST_HEAD(&rtty->http_conns);
 
     rtty->active = ev_now(rtty->loop);
+    rtty->user_active = ev_now(rtty->loop);
 
     return 0;
 }
Index: rtty-8.0.1/src/rtty.h
===================================================================
--- rtty-8.0.1.orig/src/rtty.h
+++ rtty-8.0.1/src/rtty.h
@@ -110,6 +110,7 @@ struct rtty {
     struct list_head http_conns;
     bool no_term;
     bool no_http;
+    ev_tstamp user_active;
 };
 
 int rtty_start(struct rtty *rtty);
