Index: smstools3/scripts/sendsms
===================================================================
--- smstools3.orig/scripts/sendsms
+++ smstools3/scripts/sendsms
@@ -99,6 +99,8 @@ do
   TMPFILE=`mktemp /tmp/smsd_XXXXXX`
 
   $ECHO "To: $destination" >> $TMPFILE
+  #No country code
+  [ -z "$(echo "$destination"|grep +)" ] && $ECHO "To_TOA: National" >> $TMPFILE
   [ -n "$ALPHABET" ] && $ECHO "$ALPHABET" >> $TMPFILE
   $ECHO "" >> $TMPFILE
   if [ -z "$ALPHABET" ]; then
