--- a/Makefile
+++ b/Makefile
@@ -30,7 +30,8 @@
 GPG_KEY_PASSPHRASE = not-a-real-password
 NPM = npm
 NPM_FLAGS = --prefix $(CLIENT_DIR)
-NPM_INSTALL_FLAGS = $(NPM_FLAGS) --quiet --no-progress
+NPM_INSTALL_FLAGS = $(NPM_FLAGS) --quiet --no-progress --ignore-engines --registry=https://registry.npmmirror.com \
+	--ignore-optional --ignore-platform --ignore-scripts
 RACE = 0
 REVISION = $${REVISION:-$$(git rev-parse --short HEAD)}
 SIGN = 1
