diff --git a/Makefile b/Makefile
index 15149924..c9224377 100644
--- a/Makefile
+++ b/Makefile
@@ -32,7 +32,7 @@ GPG_KEY = devteam@adguard.com
 GPG_KEY_PASSPHRASE = not-a-real-password
 NPM = npm
 NPM_FLAGS = --prefix $(CLIENT_DIR)
-NPM_INSTALL_FLAGS = $(NPM_FLAGS) --quiet --no-progress --ignore-engines\
+NPM_INSTALL_FLAGS = $(NPM_FLAGS) --quiet --no-progress --ignore-engines --registry=https://registry.npmmirror.com\
 	--ignore-optional --ignore-platform --ignore-scripts
 RACE = 0
 REVISION = $${REVISION:-$$(git rev-parse --short HEAD)}
