From ea0f0e641620e85346e77f890a1cd2cfe685f911 Mon Sep 17 00:00:00 2001
From: zengyinbin <yinbin.zeng@gl-inet.com>
Date: Fri, 21 Nov 2025 18:07:25 +0800
Subject: [PATCH] optimize flash storage usage

---
 internal/filtering/http.go | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/internal/filtering/http.go b/internal/filtering/http.go
index 94a601f..b7ba4bb 100644
--- a/internal/filtering/http.go
+++ b/internal/filtering/http.go
@@ -175,7 +175,8 @@ func (d *DNSFilter) handleFilteringRemoveURL(w http.ResponseWriter, r *http.Requ
 
 		deleted = (*filters)[delIdx]
 		p := deleted.Path(d.conf.DataDir)
-		err = os.Rename(p, p+".old")
+		// err = os.Rename(p, p+".old")
+		err = os.Remove(p)
 		if err != nil && !errors.Is(err, os.ErrNotExist) {
 			log.Error("deleting filter %d: renaming file %q: %s", deleted.ID, p, err)
 
-- 
2.34.1

