From b614be548cd0712545ab7a902f3765cd7d535c7a Mon Sep 17 00:00:00 2001
From: Jianhui Zhao <zhaojh329@gmail.com>
Date: Thu, 25 Sep 2025 17:22:18 +0800
Subject: [PATCH] fix: Fix ffi.string

Signed-off-by: Jianhui Zhao <zhaojh329@gmail.com>
---
 ffi.c | 8 +-------
 1 file changed, 1 insertion(+), 7 deletions(-)

diff --git a/ffi.c b/ffi.c
index f93bb06..2931a99 100644
--- a/ffi.c
+++ b/ffi.c
@@ -2362,14 +2362,8 @@ static int lua_ffi_string(lua_State *L)
         goto converr;
     }
 
-    switch (ct->type) {
-    case CTYPE_VOID:
-    case CTYPE_CHAR:
-    case CTYPE_UCHAR:
-        break;
-    default:
+    if (ct->type != CTYPE_CHAR)
         goto converr;
-    }
 
     if (array && array->size) {
         char *p = memchr(ptr, '\0', array->ft.size);
-- 
2.34.1

